src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
@IlyaBiryukov @AArnott I came to the same conclusion regarding where to implement a fix. The conclusion I came to is the `Dispose` method on this type is sufficiently difficult to implement in a *valid* way that it's unreasonable to expect anyone to ever use it for an actual purpose. Everyone is better off if we just remove the ability to invalidate the object altogether.